dc-commons-xml/src/main/java/de/digitalcollections/commons/xml/xpath/XPathMapper.java

```diff
@@ -70,13 +71,101 @@
+
+      Set<Field> fieldsWithMapper = ReflectionUtils.findFields(bindMapperClass, XPathRoot.class);
+      for (Field f: fieldsWithMapper) {
+        XPathRoot rootBinding = f.getAnnotation(XPathRoot.class);
+
+        // For the embedded mappers, we have to contactenate the root paths
+        String[] subMapperRootPaths = determineRootPaths(rootAnnotation, rootPaths);
+        if (rootBinding != null && rootBinding.value() != null && rootBinding.value().length == 1) {
+          for (int i=0; i<subMapperRootPaths.length; i++) {
+            subMapperRootPaths[i] = subMapperRootPaths[i] + rootBinding.value()[0];
+          }
+        }
+
+        Object subMapper = readDocument(doc, f.getType(), subMapperRootPaths);
+        ReflectionUtils.ensureAccess(f);
+        f.set(bindMapper, subMapper);
+      }
+
+      Set<Field> fieldsWithAnnotation = ReflectionUtils.findFields(bindMapperClass, XPathBinding.class);
+      for (Field f : fieldsWithAnnotation) {
+        XPathBinding fieldAnnotation = f.getAnnotation(XPathBinding.class);
+        ReflectionUtils.ensureAccess(f);
+
+        Object value;
+        Set<String> variables = xPathMapper.getVariables(fieldAnnotation.valueTemplate());
+        Set<String> expressions = new HashSet<>(Arrays.asList(fieldAnnotation.value()));
+        f.set(bindMapper, xPathMapper.evaluateXPathBinding(f.getType(), fieldAnnotation, variables, expressions));
+      }
+
+      return bindMapper;
+    } catch (Throwable t) {       // No more details available!
+      t.printStackTrace();
+      throw new RuntimeException("Cannot instantiate mapper for class " + bindMapperClass.getName() + ": " + t, t);
+    }
+  }
+
+  private static String determineDefaultRootNamespace(XPathRoot rootAnnotation) {
+    String defaultRootNamespace = "";
+    if (rootAnnotation != null && !rootAnnotation.defaultNamespace().isEmpty()) {
+      defaultRootNamespace = rootAnnotation.defaultNamespace();
+    }
+    return defaultRootNamespace;
+  }
+
+  private static String[] determineRootPaths(XPathRoot rootAnnotation, String[] rootPaths) {
+    List<String> rootPathsList = new ArrayList<>();
+    if (rootPaths != null && rootPaths.length > 0) {
+      rootPathsList = Arrays.asList(rootPaths);
+    } else if (rootAnnotation != null && rootAnnotation.value() != null && rootAnnotation.value().length > 0) {
+      rootPathsList = Arrays.asList(rootAnnotation.value());
+    }
+    return rootPathsList.toArray(new String[rootPathsList.size()]);
+  }
+
+  private static void ensureValidRootPaths(XPathRoot rootAnnotation, String[] rootPaths)
+      throws XPathMappingException {
+    // Sanity check: We must not allow two clashing root paths
+    if (rootAnnotation != null
+        && rootAnnotation.value() != null && rootAnnotation.value().length > 0
+        && rootPaths != null && rootPaths.length > 0) {
+      throw new XPathMappingException(
+          "XPathRoot is not allowed on class level and customizing field at the same time!");
```

Comment:
Die Fehlermeldung könnte aussagekräftiger sein. Das Problem ist hier dass ein Setter/Feld mit Mapper-Typ mit `@XPathRoot` annotiert ist und die Ziel-Klasse selbst auch nochmal mit `@XPathRoot` annotiert ist, oder?

Vllt lieber `XPathRoot must not be declared on classes that are embedded in other mapping classes.`

Aber: Warurm können wir die Annotation auf der Ziel-Klasse nicht einfach ignorieren? Würde die Wiederverwendbarkeit von Mapping-Klassen erhöhen für kaum Mehraufwand.